aten/src/ATen/native/cuda/GridSampler.cu

```diff
@@ -0,0 +1,782 @@
+			if (padding_mode == GridSamplerPadding::Border) {
+			  // clip coordinates to image borders
+			  ix_nw_cl = clip_coordinates(ix_nw, inp_W);
+			  iy_nw_cl = clip_coordinates(iy_nw, inp_H);
+			  ix_ne_cl = clip_coordinates(ix_ne, inp_W);
+			  iy_ne_cl = clip_coordinates(iy_ne, inp_H);
+			  ix_sw_cl = clip_coordinates(ix_sw, inp_W);
+			  iy_sw_cl = clip_coordinates(iy_sw, inp_H);
+			  ix_se_cl = clip_coordinates(ix_se, inp_W);
+			  iy_se_cl = clip_coordinates(iy_se, inp_H);
+			} else {
+	      ix_nw_cl = ix_nw;
+	      iy_nw_cl = iy_nw;
+	      ix_ne_cl = ix_ne;
+	      iy_ne_cl = iy_ne;
+	      ix_sw_cl = ix_sw;
+	      iy_sw_cl = iy_sw;
+	      ix_se_cl = ix_se;
+	      iy_se_cl = iy_se;
+			}
+
+      scalar_t gix = static_cast<scalar_t>(0), giy = static_cast<scalar_t>(0);
+      scalar_t *gOut_ptr_NCHW = grad_output.data + n * gOut_sN + h * gOut_sH + w * gOut_sW;
+      scalar_t *gInp_ptr_NC = grad_input.data + n * gInp_sN;
+      scalar_t *inp_ptr_NC = input.data + n * inp_sN;
+	    for (int c = 0; c < C; ++c, inp_ptr_NC += inp_sC, gInp_ptr_NC += gInp_sC, gOut_ptr_NCHW += gOut_sC) {
+        scalar_t gOut = *gOut_ptr_NCHW;
+
+	      // calculate and set grad_input
+	      safe_add_2d(gInp_ptr_NC, iy_nw_cl, ix_nw_cl, gInp_sH, gInp_sW, inp_H, inp_W, nw * gOut);
+	      safe_add_2d(gInp_ptr_NC, iy_ne_cl, ix_ne_cl, gInp_sH, gInp_sW, inp_H, inp_W, ne * gOut);
+	      safe_add_2d(gInp_ptr_NC, iy_sw_cl, ix_sw_cl, gInp_sH, gInp_sW, inp_H, inp_W, sw * gOut);
+	      safe_add_2d(gInp_ptr_NC, iy_se_cl, ix_se_cl, gInp_sH, gInp_sW, inp_H, inp_W, se * gOut);
+
+	      // calculate grad_grid
+	      if (padding_mode != GridSamplerPadding::Zeros || within_bounds_2d(iy_nw_cl, ix_nw_cl, inp_H, inp_W)) {
+	      	scalar_t nw_val = inp_ptr_NC[iy_nw_cl * inp_sH + ix_nw_cl * inp_sW];
+	      	gix -= nw_val * (iy_se - iy) * gOut;
+      		giy -= nw_val * (ix_se - ix) * gOut;
+	      }
+	      if (padding_mode != GridSamplerPadding::Zeros || within_bounds_2d(iy_ne_cl, ix_ne_cl, inp_H, inp_W)) {
+	      	scalar_t ne_val = inp_ptr_NC[iy_ne_cl * inp_sH + ix_ne_cl * inp_sW];
+	      	gix += ne_val * (iy_sw - iy) * gOut;
+      		giy -= ne_val * (ix - ix_sw) * gOut;
+	      }
+	      if (padding_mode != GridSamplerPadding::Zeros || within_bounds_2d(iy_sw_cl, ix_sw_cl, inp_H, inp_W)) {
+	      	scalar_t sw_val = inp_ptr_NC[iy_sw_cl * inp_sH + ix_sw_cl * inp_sW];
+	      	gix -= sw_val * (iy - iy_ne) * gOut;
+      		giy += sw_val * (ix_ne - ix) * gOut;
+	      }
+	      if (padding_mode != GridSamplerPadding::Zeros || within_bounds_2d(iy_se_cl, ix_se_cl, inp_H, inp_W)) {
+	      	scalar_t se_val = inp_ptr_NC[iy_se_cl * inp_sH + ix_se_cl * inp_sW];
+	        gix += se_val * (iy - iy_nw) * gOut;
+	        giy += se_val * (ix - ix_nw) * gOut;
+	      }
+	    }
+
+	    // un-normalize grad_grid values back to [-1, 1] constraints
+	    gix = gix * (inp_W - 1.f) / 2;
+	    giy = giy * (inp_H - 1.f) / 2;
+
+      // assuming grad_grid is contiguous
```

Comment:
I don't understand.

In the original code, the gradGrid update looks like:

```
    Dtype gix_old = gradGrid[n][h][w][0];
    Dtype giy_old = gradGrid[n][h][w][1];

    gradGrid[n][h][w][0] = gix_old + gix;
    gradGrid[n][h][w][1] = giy_old + giy;
```

So, I see two major changes: (1) the access pattern on gradGrid is different (because you transposed N and W) and (2) you're no longer accumulating into grad grid; instead you're writing to it directly.

I can see (1) being good for perf (though I haven't thought enough about it to know if it makes sense). But (2) doesn't really care about the contiguity of gradGrid, all that is necessary is that we only handle any given N/H/W combination once, which I think was true for the old code too.

So, is the use of "contiguity" of gradGrid here just to make sure you get to do contiguous writes into gradGrid? If so, say so!

Reply:
(And, I guess, you get to avoid some index computation, though I think the optimizer should be able to figure it out, since once it inlines the values of n/h/w it should be able to do the required arithmetic simplifications).

Reply:
Yeah, (2) doesn't require contiguity. In the THUCNN code, `grad_grid` is initialized as all zeros tensor, and each element is accumulated exactly once, so I changed to directly writing.

The contiguity assumption is used in computing `gGrid_ptr_NHW` and just assigning `gGrid_ptr_NHW[0]` and `gGrid_ptr_NHW[1]` (assuming last dimension has stride 1). 

I'll add some comments in code.